File: leetcode_notebooks/quicksort.py
'''
Variant: Algo2 Uni Goethe
'''
#import random
#A = random.sampAe(range(0,101), 9)
A = [42, 32, 96, 23, 13, 24, 94, 52, 71]
# A = [2,8,7,1,3,5,6,4]
p = 0
r = len(A)- 1

def swap(i, j):
    temp = A[i]
    A[i] = A[j]
    A[j] = temp

def pivot(links, rechts):
    print(links, rechts)
    return rechts

def partition(p, links, rechts):
    alpha = A[p]
    l = links
    r = rechts
    swap(rechts, p)
    while l < r:
        while A[l] <= alpha:
            l += 1
        while A[r] >= alpha and r >= 0:
            r -= 1
        if l < r:
            swap(l,r)
    swap(rechts, l)
    return l


def quicksort(links, rechts):
    if links < rechts:
        print(A[links:rechts+1])
        p = pivot(links, rechts)
        i = partition(p, links, rechts)
        quicksort(links, i-1)
        quicksort(i+1, rechts)
quicksort(0, len(A)-1)
print(A)

File: leetcode_notebooks/quicksort.cpp
#include <iostream>
using namespace std;

int A[] = { 5,4,3,2,1 };

void printArray() {
    for (int i : A) {
        cout << i << " ";
    }
    cout << "\n";
}

void swap(int i, int j) {
    int temp = A[i];
    A[i] = A[j];
    A[j] = temp;
}

int pivot(int links, int rechts) {
    return(rechts);
}

int partition(int p, int links, int rechts) {
    int alpha, l, r;
    alpha = A[p];
    l = links - 1;
    A[l] = INT8_MIN;
    r = rechts;
    swap(r,p);
    cout << "p " << p << " l " << l << " r " << r << "\n";
    do {
        cout << "Comp l " << l << " " << A[l] << " " << alpha << "\n";
        while (A[l] <= alpha){
            l++;
        }
        cout << "Comp l " << r << " " << A[r]  << " " << alpha << "\n";
        while (A[r] >= alpha){
            r--;
        }
        if (l < r){
            swap(l,r);
        }
    } while (l < r);
    swap(rechts, l);
    return l;
}

void quicksort(int links, int rechts) {
    int p, i;
    if (links < rechts) {
        cout << "asd";
        p = pivot(links, rechts);
        i = partition(p, links, rechts);
        cout << "Pivot: " << i << " Value: " << A[i] << "\n";
        quicksort(links, i-1);
        quicksort(i+1, rechts);
        printArray();
    }
}

int main() {
    printArray();
    quicksort(0,4);
}